File: crates/aip-core/src/document.rs
//! DID Document structure and signing.
//!
//! A DID Document contains the public keys and service endpoints for an agent.
//! Documents are self-signed to prevent tampering.

use crate::{signing, Did, Error, Result, RootKey};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// A verification method (public key) in a DID Document.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationMethod {
    /// Full ID of this key (e.g., "did:key:z6Mk...#root")
    pub id: String,
    /// Key type
    #[serde(rename = "type")]
    pub type_: String,
    /// Controller DID
    pub controller: String,
    /// Public key in multibase format
    pub public_key_multibase: String,
}

/// A service endpoint in a DID Document.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Service {
    /// Service ID (e.g., "did:key:z6Mk...#handshake")
    pub id: String,
    /// Service type
    #[serde(rename = "type")]
    pub type_: String,
    /// Service endpoint URL
    pub service_endpoint: String,
}

/// Proof attached to a DID Document.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentProof {
    /// Proof type
    #[serde(rename = "type")]
    pub type_: String,
    /// When the proof was created
    pub created: DateTime<Utc>,
    /// Which key was used to sign
    pub verification_method: String,
    /// The signature value (base64)
    pub proof_value: String,
}

/// A DID Document describing an agent's keys and services.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DidDocument {
    /// JSON-LD context
    #[serde(rename = "@context")]
    pub context: Vec<String>,
    /// The DID this document describes
    pub id: String,
    /// Controller of this DID (usually self)
    pub controller: String,
    /// Verification methods (public keys)
    pub verification_method: Vec<VerificationMethod>,
    /// Keys that can authenticate as this DID
    pub authentication: Vec<String>,
    /// Keys that can make assertions
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assertion_method: Option<Vec<String>>,
    /// Service endpoints
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service: Option<Vec<Service>>,
    /// When this document was created
    pub created: DateTime<Utc>,
    /// When this document was last updated
    pub updated: DateTime<Utc>,
    /// Proof of authenticity
    #[serde(skip_serializing_if = "Option::is_none")]
    pub proof: Option<DocumentProof>,
}

impl DidDocument {
    /// Create a new unsigned DID Document for a root key.
    pub fn new(root_key: &RootKey) -> Self {
        let did = root_key.did();
        let did_str = did.to_string();
        let now = Utc::now();

        // Encode public key as multibase (z = base58btc)
        let pubkey_multibase = format!("z{}", did.key_id());

        Self {
            context: vec![
                "https://www.w3.org/ns/did/v1".to_string(),
                "https://w3id.org/security/suites/ed25519-2020/v1".to_string(),
            ],
            id: did_str.clone(),
            controller: did_str.clone(),
            verification_method: vec![VerificationMethod {
                id: format!("{}#root", did_str),
                type_: "Ed25519VerificationKey2020".to_string(),
                controller: did_str.clone(),
                public_key_multibase: pubkey_multibase,
            }],
            authentication: vec![format!("{}#root", did_str)],
            assertion_method: Some(vec![format!("{}#root", did_str)]),
            service: None,
            created: now,
            updated: now,
            proof: None,
        }
    }

    /// Add a service endpoint.
    pub fn add_service(mut self, id: &str, type_: &str, endpoint: &str) -> Self {
        let service = Service {
            id: format!("{}#{}", self.id, id),
            type_: type_.to_string(),
            service_endpoint: endpoint.to_string(),
        };

        match &mut self.service {
            Some(services) => services.push(service),
            None => self.service = Some(vec![service]),
        }

        self.updated = Utc::now();
        self
    }

    /// Add the handshake service endpoint.
    pub fn with_handshake_endpoint(self, endpoint: &str) -> Self {
        self.add_service("handshake", "AIPHandshake", endpoint)
    }

    /// Sign this document with the root key.
    pub fn sign(mut self, root_key: &RootKey) -> Result<Self> {
        // Clear any existing proof before signing
        self.proof = None;
        self.updated = Utc::now();

        // Canonicalize and sign
        let canonical = signing::canonicalize(&self)?;
        let signature = root_key.sign(&canonical);
        let sig_b64 = base64::Engine::encode(
            &base64::engine::general_purpose::STANDARD,
            signature.to_bytes(),
        );

        self.proof = Some(DocumentProof {
            type_: "Ed25519Signature2020".to_string(),
            created: Utc::now(),
            verification_method: format!("{}#root", self.id),
            proof_value: sig_b64,
        });

        Ok(self)
    }

    /// Verify this document's signature.
    pub fn verify(&self) -> Result<()> {
        let proof = self.proof.as_ref().ok_or(Error::InvalidSignature)?;

        // Parse the DID to get the expected public key
        let did: Did = self.id.parse()?;
        let public_key = did.public_key()?;

        // Create unsigned version for verification
        let mut unsigned = self.clone();
        unsigned.proof = None;

        let canonical = signing::canonicalize(&unsigned)?;

        // Decode and verify signature
        let sig_bytes = base64::Engine::decode(
            &base64::engine::general_purpose::STANDARD,
            &proof.proof_value,
        )
        .map_err(|_| Error::InvalidSignature)?;

        let signature = ed25519_dalek::Signature::from_bytes(
            &sig_bytes.try_into().map_err(|_| Error::InvalidSignature)?,
        );

        crate::keys::verify(&public_key, &canonical, &signature)?;

        Ok(())
    }

    /// Get the handshake endpoint if present.
    pub fn handshake_endpoint(&self) -> Option<&str> {
        self.service.as_ref()?.iter().find_map(|s| {
            if s.type_ == "AIPHandshake" {
                Some(s.service_endpoint.as_str())
            } else {
                None
            }
        })
    }

    /// Get the DID this document describes.
    pub fn did(&self) -> Result<Did> {
        self.id.parse()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_document_creation() {
        let key = RootKey::generate();
        let doc = DidDocument::new(&key);

        assert!(doc.id.starts_with("did:key:z6Mk"));
        assert_eq!(doc.verification_method.len(), 1);
        assert!(doc.proof.is_none());
    }

    #[test]
    fn test_document_signing() {
        let key = RootKey::generate();
        let doc = DidDocument::new(&key)
            .with_handshake_endpoint("https://example.com/handshake")
            .sign(&key)
            .unwrap();

        assert!(doc.proof.is_some());
        doc.verify().unwrap();
    }

    #[test]
    fn test_document_tamper_detection() {
        let key = RootKey::generate();
        let mut doc = DidDocument::new(&key).sign(&key).unwrap();

        // Tamper with the document
        doc.controller = "did:key:z6MkATTACKER".to_string();

        // Verification should fail
        assert!(doc.verify().is_err());
    }

    #[test]
    fn test_handshake_endpoint() {
        let key = RootKey::generate();
        let doc = DidDocument::new(&key).with_handshake_endpoint("https://example.com/hs");

        assert_eq!(doc.handshake_endpoint(), Some("https://example.com/hs"));
    }
}

File: sdk/python/agent_id/document.py
"""
DID Document handling for Agent Identity Protocol.

A DID Document describes an agent's identity, including:
- Verification methods (public keys)
- Service endpoints (how to reach the agent)
- Authentication methods
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from agent_id.did import Did
from agent_id.keys import RootKey
from agent_id.signing import canonicalize, sign_message, verify_message


@dataclass
class VerificationMethod:
    """A verification method (public key) in a DID Document."""
    
    id: str
    type: str
    controller: str
    public_key_multibase: str


@dataclass
class Service:
    """A service endpoint in a DID Document."""
    
    id: str
    type: str
    service_endpoint: str


@dataclass
class DidDocument:
    """
    A DID Document describing an agent's identity.
    
    Example:
        >>> key = RootKey.generate()
        >>> doc = DidDocument.new(key)
        >>> doc = doc.with_handshake_endpoint("https://agent.example/aip")
        >>> signed_doc = doc.sign(key)
    """
    
    did: Did
    verification_methods: List[VerificationMethod] = field(default_factory=list)
    authentication: List[str] = field(default_factory=list)
    assertion_method: List[str] = field(default_factory=list)
    services: List[Service] = field(default_factory=list)
    created: Optional[datetime] = None
    updated: Optional[datetime] = None
    proof: Optional[Dict[str, Any]] = None
    
    @classmethod
    def new(cls, root_key: RootKey) -> DidDocument:
        """
        Create a new DID Document for a root key.
        
        Args:
            root_key: The agent's root key
            
        Returns:
            An unsigned DID Document
        """
        import base58
        
        did = root_key.did
        key_id = f"{did}#root"
        
        # Encode public key as multibase (z = base58btc)
        public_key_multibase = "z" + base58.b58encode(bytes(root_key.verify_key)).decode("ascii")
        
        verification_method = VerificationMethod(
            id=key_id,
            type="Ed25519VerificationKey2020",
            controller=str(did),
            public_key_multibase=public_key_multibase,
        )
        
        now = datetime.now(timezone.utc)
        
        return cls(
            did=did,
            verification_methods=[verification_method],
            authentication=[key_id],
            assertion_method=[key_id],
            created=now,
            updated=now,
        )
    
    def with_handshake_endpoint(self, endpoint: str) -> DidDocument:
        """
        Add a handshake service endpoint.
        
        Args:
            endpoint: The URL for AIP handshake
            
        Returns:
            A new DidDocument with the service added
        """
        service = Service(
            id=f"{self.did}#handshake",
            type="AIPHandshake",
            service_endpoint=endpoint,
        )
        
        return DidDocument(
            did=self.did,
            verification_methods=self.verification_methods,
            authentication=self.authentication,
            assertion_method=self.assertion_method,
            services=[*self.services, service],
            created=self.created,
            updated=datetime.now(timezone.utc),
            proof=None,  # Clear proof since document changed
        )
    
    def with_service(self, service_id: str, service_type: str, endpoint: str) -> DidDocument:
        """
        Add a custom service endpoint.
        
        Args:
            service_id: Short ID for the service (will be prefixed with DID)
            service_type: The service type
            endpoint: The service URL
            
        Returns:
            A new DidDocument with the service added
        """
        service = Service(
            id=f"{self.did}#{service_id}",
            type=service_type,
            service_endpoint=endpoint,
        )
        
        return DidDocument(
            did=self.did,
            verification_methods=self.verification_methods,
            authentication=self.authentication,
            assertion_method=self.assertion_method,
            services=[*self.services, service],
            created=self.created,
            updated=datetime.now(timezone.utc),
            proof=None,
        )
    
    def to_dict(self, include_proof: bool = True) -> Dict[str, Any]:
        """
        Convert to a JSON-serializable dictionary.
        
        Args:
            include_proof: Whether to include the proof field
            
        Returns:
            DID Document as a dictionary
        """
        doc: Dict[str, Any] = {
            "@context": [
                "https://www.w3.org/ns/did/v1",
                "https://aip.network/v1",
            ],
            "id": str(self.did),
            "controller": str(self.did),
            "verificationMethod": [
                {
                    "id": vm.id,
                    "type": vm.type,
                    "controller": vm.controller,
                    "publicKeyMultibase": vm.public_key_multibase,
                }
                for vm in self.verification_methods
            ],
            "authentication": self.authentication,
            "assertionMethod": self.assertion_method,
        }
        
        if self.services:
            doc["service"] = [
                {
                    "id": s.id,
                    "type": s.type,
                    "serviceEndpoint": s.service_endpoint,
                }
                for s in self.services
            ]
        
        if self.created:
            doc["created"] = self.created.isoformat().replace("+00:00", "Z")
        
        if self.updated:
            doc["updated"] = self.updated.isoformat().replace("+00:00", "Z")
        
        if include_proof and self.proof:
            doc["proof"] = self.proof
        
        return doc
    
    def sign(self, key: RootKey) -> DidDocument:
        """
        Sign this DID Document.
        
        Args:
            key: The root key to sign with (must match document DID)
            
        Returns:
            A new DidDocument with proof attached
            
        Raises:
            ValueError: If key doesn't match document DID
        """
        if key.did != self.did:
            raise ValueError(f"Key DID {key.did} doesn't match document DID {self.did}")
        
        # Get document without proof for signing
        doc_dict = self.to_dict(include_proof=False)
        
        signature = sign_message(doc_dict, key)
        
        proof = {
            "type": "Ed25519Signature2020",
            "created": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "verificationMethod": f"{self.did}#root",
            "proofValue": signature,
        }
        
        return DidDocument(
            did=self.did,
            verification_methods=self.verification_methods,
            authentication=self.authentication,
            assertion_method=self.assertion_method,
            services=self.services,
            created=self.created,
            updated=self.updated,
            proof=proof,
        )
    
    def verify(self) -> bool:
        """
        Verify the document's signature.
        
        Returns:
            True if signature is valid
            
        Raises:
            ValueError: If document has no proof
        """
        if not self.proof:
            raise ValueError("Document has no proof to verify")
        
        # Get document without proof
        doc_dict = self.to_dict(include_proof=False)
        
        signature_b64 = self.proof.get("proofValue", "")
        
        return verify_message(doc_dict, signature_b64, self.did.public_key)
    
    def __repr__(self) -> str:
        signed = "signed" if self.proof else "unsigned"
        return f"DidDocument({self.did}, {signed})"
